### PR TITLE
Move webhook setup before manager start

### PR DESCRIPTION
**Pull request by yiqigao217**
According to
https://github.com/kubernetes-sigs/controller-runtime/issues/1148,
there's a race condition between manager start and webhook register.
This commit moved webhook setup before manager start.

Tested on GKE cluster. The nil pointer error was not detected for 20
times.

### CODE REVIEW

File: incubator/hnc/cmd/manager/main.go [hunk 1]
```diff
@@ -153,32 +153,8 @@ func main() {
...
-	setupLog.Info("Waiting for certificate generation to complete")
-	// Block until the setup finishes.
-	<-setupFinished
-
-	if testLog {
-		stats.StartLoggingActivity()
-	}
-
-	// Create all reconciling controllers
 	f := forest.NewForest()
```
Comment (adrianludwin) [inline incubator/hnc/cmd/manager/main.go:156]: This is a very significant line so I might comment it :)

File: incubator/hnc/cmd/manager/main.go [hunk 2]
```diff
@@ -153,32 +153,8 @@ func main() {
...
-	setupLog.Info("Creating controllers", "maxReconciles", maxReconciles)
-	removeOldCRDVersion := true
-	if err := reconcilers.Create(mgr, f, maxReconciles, removeOldCRDVersion); err != nil {
-		setupLog.Error(err, "cannot create controllers")
-		os.Exit(1)
-	}
+	go startControllers(mgr, f, setupFinished)
```
Comment (adrianludwin) [inline incubator/hnc/cmd/manager/main.go:157]: While it _probably_ doesn't really matter, I might move this to be the last statement before starting the manager. This reduces the amount of possible parallelism.

File: incubator/hnc/cmd/manager/main.go [hunk 3]
```diff
@@ -153,32 +153,8 @@ func main() {
...
 
 	// Create all validating admission controllers.
```
Comment (adrianludwin) [inline incubator/hnc/cmd/manager/main.go:159]: Maybe add a comment here about the bug we're working around. I know it won't exist in the future, but it's nice to say why we're starting the webhooks before the reconcilers.

**Review by adrianludwin — COMMENTED** [review]
lgtm with minor nits. You don't need to test another 20 times after making my suggested changes, once will be fine :)

File: incubator/hnc/cmd/manager/main.go (hunk 2 above)
Comment (yiqigao217) [reply incubator/hnc/cmd/manager/main.go:157]: sg

**Comment by yiqigao217** [comment]
> lgtm with minor nits. You don't need to test another 20 times after making my suggested changes, once will be fine :)

Updated. BTW tested 20 times because last time it took about 20 times to reproduce the race condition.

**Review by adrianludwin — COMMENTED** [review]
/lgtm
/approve

@rjbez17 this is a fairly straightforward fix, lmk if you have any concerns about it but I'm just approving for now.

**Comment by yiqigao217** [comment]
/retest

looking the object reconciler issue now.

**Pull request merged**